readthedocs/readthedocs.org
# Search: use multi-fields for Wildcard queries

Wildcard queries are slow (on .org it returns 502, on .com since the db
is small it works just fine).

https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html#query-dsl-allow-expensive-queries

These type of queries can be optimized by using the Wildcard field
https://www.elastic.co/guide/en/elasticsearch/reference/current/keyword.html#wildcard-field-type.

Currently that field isn't implemented in the ES dependencies,
but is just a subclass of the Keyword field (I'll see if I can send a PR
upstream).

As we still want to make use of the SimpleString queries,
I'm using the multi-fields feature.

This change is editing the index,
so we need to rebuild the index and re-index.

To test it locally:

- `inv docker.manage 'search_index --rebuild'`.
- `inv docker.manage reindex_elasticsearch`.
- Enable the `DEFAULT_TO_FUZZY_SEARCH` feature flag on a project.

I'm not sure how to write tests for this one,
and we can't test if this is really fast in production...
But testing it locally works and gives better results for both, sections and
domains!

This is based on https://github.com/readthedocs/readthedocs.org/pull/7582 since the wildcard field was added in ES 7.x.
As we are going to re-index when deploying 7.x, I think we can deploy this change together with that (otherwise we'll need to do the re-index again).

One example where you can see better results from domains (with the wildcard search  there is a match in the title instead of matching the content)

![Screenshot_2020-10-28 Read the Docs Documentation Simplified — Read the Docs 5 6 1 documentation(1)](https://user-images.githubusercontent.com/4975310/97481635-19109900-1923-11eb-81d8-384957e589bf.png)
![Screenshot_2020-10-28 Read the Docs Documentation Simplified — Read the Docs 5 6 1 documentation](https://user-images.githubusercontent.com/4975310/97481637-1a41c600-1923-11eb-8104-af9606a53f8c.png)
